### PR TITLE
support setting periodSeconds and failureThreashold in the Keyclock CR

### LINKED ISSUE #21995: Configurable probes in the Operator
### Description

The probes that are set up by the Operator should be configurable. We should expose PeriodSeconds and FailureThreshold separately for startup, liveness and readiness probes in the Keycloak CR.

### Discussion

https://github.com/keycloak/keycloak/pull/21884#pullrequestreview-1547156709

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by KyriosGN0**
Signed-off-by: AvivGuiser <avivguiser@gmail.com>

fixes #21995

### CODE REVIEW

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java [hunk 1]
```diff
@@ -135,6 +136,20 @@ public class KeycloakSpec {
...
+    private ProbeSpec readinessProbeSpec;
+
+
+    @JsonProperty("livenessProbe")
+    @JsonPropertyDescription("Configuration for liveness probe")
+    private ProbeSpec livenessProbeSpec;
+
+    @JsonProperty("startupProbe")
+    @JsonPropertyDescription("Configuration for startup probe")
+    private ProbeSpec startupProbeSpec;
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java]: It'd be good to have the defaults mentioned at least in the desc.

**Review by vmuzikar — COMMENTED** [review]
@KyriosGN0 Thank you for the PR!

It'd be good to also have a mention about this in the advanced Operator guide.

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java (hunk 1 above)
Comment (KyriosGN0) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java]: added

**Comment by KyriosGN0** [comment]
> @KyriosGN0 Thank you for the PR!
> 
> It'd be good to also have a mention about this in the advanced Operator guide.

i have added a section about it

File: operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakDeploymentTest.java [hunk 2]
```diff
@@ -728,4 +729,50 @@ private void handleFakeImagePullSecretCreation(Keycloak keycloakCR,
         LocalObjectReference localObjRefAsSecretTmp = new LocalObjectReferenceBuilder().withName(imagePullSecret.getMetadata().getName()).build();
         keycloakCR.getSpec().setImagePullSecrets(Collections.singletonList(localObjRefAsSecretTmp));
     }
+
+    @Test
+    public void testProbeConfig(){
```
Comment (shawkins) [inline operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakDeploymentTest.java]: I would prefer this test as a unit test (see PodTemplateTest) as we're not testing a boundary condition such as a probe setting would make the deployment fail. We can avoid adding test time waiting for the keycloak instance to come up and get the same coverage.

File: operator/src/test/resources/example-probe.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+apiVersion: k8s.keycloak.org/v2alpha1
```
Comment (shawkins) [inline operator/src/test/resources/example-probe.yaml]: Are you using this in a test? If not, it probably should be incorporated into something like test-serialization-keycloak-cr.yaml. 

These files are just to support tests and won't be referenced as actual examples.

**Review by shawkins — COMMENTED** [review]
Thank you @KyriosGN0 overall everything looks good. Just a couple of comments.

File: operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakDeploymentTest.java (hunk 2 above)
Comment (KyriosGN0) [reply operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakDeploymentTest.java]: moved it to a unittest

File: operator/src/test/resources/example-probe.yaml (hunk 3 above)
Comment (KyriosGN0) [reply operator/src/test/resources/example-probe.yaml]: moved to the test-serialization-keycloak-cr.yaml file

File: operator/src/test/resources/test-serialization-keycloak-cr.yml [hunk 4]
```diff
@@ -32,6 +32,15 @@ spec:
     annotations:
       myAnnotation: myValue
       anotherAnnotation: anotherValue
+  readinessProbe:
```
Comment (shawkins) [inline operator/src/test/resources/test-serialization-keycloak-cr.yml:35]: One last thing - can you add some assertions for these to CRSerializationTest?

Comment (KyriosGN0) [reply operator/src/test/resources/test-serialization-keycloak-cr.yml:35]: @shawkins done

**Review by shawkins — DISMISSED** [review]
LGTM - thank you @KyriosGN0

**Comment by KyriosGN0** [comment]
@vmuzikar @shawkins anything i can do about the CI failing?

**Comment by shawkins** [comment]
@KyriosGN0 If this is happening reproducibly, then it will take some further investigation to trap reason why it's failing. There's nothing in the logs that help determine why the instance isn't becoming ready - we do try to obtain the current or previous pod log from what is failing, but nothing is showing up.

**Comment by KyriosGN0** [comment]
locally this specific test passes

**Comment by shawkins** [comment]
> locally this specific test passes

I reran the tests and it's passing now as expected. We'll need to address the flakiness of that test in a separate issue, but I don't recall it being something that commonly fails.

**Comment by KyriosGN0** [comment]
@shawkins @vmuzikar i don't mean to be nagging, i fixed the conflict i had and then i noticed that the test i wrote ws confusing (i always built readiness probe)

**Comment by KyriosGN0** [comment]
@shawkins @vmuzikar can i get a retest? Thanks in advance

**Comment by shawkins** [comment]
> @shawkins @vmuzikar can i get a retest? Thanks in advance

Hopefully another pr will stablize the olm tests soon, we'll then pick that up here and let it run again.

**Review by shawkins — DISMISSED** [review]
LGTM, thanks @KyriosGN0 

@vmuzikar do you want to give it a look?

**Review by vmuzikar — APPROVED** [review]
LGTM, thank you.

File: docs/guides/operator/advanced-configuration.adoc [hunk 5]
```diff
@@ -149,6 +149,28 @@ spec:
...
+    failureThreshold: 5
+  livenessProbe:
+    periodSeconds: 20
+    failureThreshold: 5
+  startupProbe:
+    periodSeconds: 20
+    failureThreshold: 5
+----
+
 ===== Probe Timeouts
```
Comment (vmuzikar) [inline docs/guides/operator/advanced-configuration.adoc]: Actually, we might want to remove the next section "Probe Timeouts" for using pod templated. It's superseded by this new one.

**Review by vmuzikar — CHANGES_REQUESTED** [review]

File: docs/guides/operator/advanced-configuration.adoc (hunk 5 above)
Comment (KyriosGN0) [reply docs/guides/operator/advanced-configuration.adoc]: updated docs, thanks!

**Review by vmuzikar — APPROVED** [review]
@KyriosGN0 Thanks.
